tests/unit/roommanager.spec.js

```diff
@@ -0,0 +1,129 @@
+    InfoExtract.getVideoInfo = jest.fn().mockReturnValue(new Promise(resolve => resolve({ service: "youtube", id: "I3O9J02G67I", length: 10 })));
+    expect(roommanager.rooms).toBeDefined();
+    expect(roommanager.rooms.test).toBeDefined();
+    expect(roommanager.rooms.test.name).toBeDefined();
+    expect(roommanager.rooms.test.name.length).toBeGreaterThan(0);
+    expect(roommanager.rooms.test.queue.length).toEqual(0);
+    expect(roommanager.rooms.test.currentSource).toEqual({});
+
+    expect(roommanager.addToQueue(roommanager.rooms.test.name, { url: "http://youtube.com/watch?v=I3O9J02G67I" })).resolves.toBe(true);
+
+    expect(roommanager.rooms.test).toBeDefined();
+    expect(InfoExtract.getVideoInfo).toBeCalledWith("youtube", "I3O9J02G67I");
+    expect(roommanager.rooms.test.queue.length).toEqual(0);
+    // Make sure that any async functions waiting have finished before checking currentSource
+    setImmediate(() => {
+      expect(roommanager.rooms.test.currentSource).toEqual({ service: "youtube", id: "I3O9J02G67I", length: 10 });
+      done();
+    });
+  });
+
+  it('should add a video to the queue with service and id provided, and because no video is playing, move it into currentSource', done => {
+    InfoExtract.getVideoInfo = jest.fn().mockReturnValue(new Promise(resolve => resolve({ service: "youtube", id: "I3O9J02G67I", length: 10 })));
+    expect(roommanager.rooms).toBeDefined();
+    expect(roommanager.rooms.test).toBeDefined();
+    expect(roommanager.rooms.test.name).toBeDefined();
+    expect(roommanager.rooms.test.name.length).toBeGreaterThan(0);
+    expect(roommanager.rooms.test.queue.length).toEqual(0);
+    expect(roommanager.rooms.test.currentSource).toEqual({});
+
+    expect(roommanager.addToQueue(roommanager.rooms.test.name, { service: "youtube", id: "I3O9J02G67I" })).resolves.toBe(true);
+
+    expect(roommanager.rooms.test).toBeDefined();
+    expect(InfoExtract.getVideoInfo).toBeCalledWith("youtube", "I3O9J02G67I");
+    expect(roommanager.rooms.test.queue.length).toEqual(0);
+    // Make sure that any async functions waiting have finished before checking currentSource
+    setImmediate(() => {
+      expect(roommanager.rooms.test.currentSource).toEqual({ service: "youtube", id: "I3O9J02G67I", length: 10 });
+      done();
+    });
+  });
+
+  it('should add a video to the queue with service and id provided, and because a video is playing, leave it in the queue', done => {
+    InfoExtract.getVideoInfo = jest.fn().mockReturnValue(new Promise(resolve => resolve({ service: "youtube", id: "I3O9J02G67I", length: 10 })));
+    expect(roommanager.rooms).toBeDefined();
+    expect(roommanager.rooms.test).toBeDefined();
+    expect(roommanager.rooms.test.name).toBeDefined();
+    expect(roommanager.rooms.test.name.length).toBeGreaterThan(0);
+    expect(roommanager.rooms.test.queue.length).toEqual(0);
```

Comment:
I don't usually write tests but is there a way to reduce this repetition?

Reply:
Maybe. Since I'm rewriting the room manager, this will probably get changed anyway.